### ISSUE TITLE
Question abut performance

### ISSUE THREAD

**Issue by piegus** [select2/select2#3134]
Hey guys. Greate job on select2 btw.
I would like to ask what is the perfomarnce of new select2 v 4.0?
Is it faster to previous version? Have somebody tried to measure it?

**Comment by kevin-brown** [comment]
> Is it faster to previous version?

Much faster, I put quite a bit of effort into making sure that the major performance issues were removed. You can read a bit about it [in the blog post I wrote](http://blog.kevinbrown.in/select2/2014/12/15/jquery-js-performance.html).

> Have somebody tried to measure it?

I don't have any recent measurements, but at the time I was able to drop querying 10,000 `<option>` elements from 2.5 seconds to just under half a second on the first query. It's much quicker after the first query (under 100ms) because we do quite a bit of caching.

If someone is interested in contributing performance tests, that'd be helpful to allow us to quantify the improvements.

**Comment by vol7ron** [comment]
Nice work Kevin.  Is this only for AJAX requests?  Is searching any faster if constructing from HTML elements, or JSON (non-AJAX requests)?

**Comment by kevin-brown** [comment]
> Is this only for AJAX requests?

AJAX requests weren't causing most of the performance issues in Select2 3.x. While there were some performance issues relating to caching data objects, the performance was already pretty good.

> Is searching any faster if constructing from HTML elements

It's much faster if working from a `<select>` with lots (10,000+) of elements. Right now there is a slight delay for the first search (when the dropdown opens) which has to due with the initial caching, but it shouldn't be noticeable and it should only happen once.

> or JSON (non-AJAX requests)?

This is in the same boat as AJAX requests. The only difference is that we automatically convert `data`-provided queries into standard `<option>` elements when Select2 is initialized.

**Comment by j-ro** [comment]
Hey Kevin, thanks for the update! Indeed, it is a fair bit faster. Still having some bottlenecks though, check out this fiddle with 10k items:

https://jsfiddle.net/f57fd7mj/4/

The bottleneck appears to be opening the menu -- selecting items, searching, all that is great, but anything that opens that dropdown is very slow, 3 seconds or so on my machine.

Any ideas on how I might be able to speed that up? Seems like that could potentially be built and cached on load, so opening doesn't take so long.

**Comment by j-ro** [comment]
Note that this answer really helps for big lists:

http://stackoverflow.com/a/25466453/1974848

Would be cool to see that kind of internal lazy loading built in.

**Comment by molily** [comment]
Hello! I’m profiling with this example with a lot of options, especially with `ArrayAdapter`  (`data`) and `templateResult`:
http://codepen.io/molily/pen/EjEZKQ
I’m using the Timeline and Profiles of the Chrome Dev Tools. There are a some things I noticed and I’d like to mention, hopefully that’s helpful for further perf optimization.

1. four calls of `SelectAdapter.prototype.current`. This is costly because of `.find(':selected')`. All selected options need to be walked, plus `parentNode.selectedIndex` is read each time ([this is how Sizzle works](https://github.com/jquery/sizzle/blob/2.2.0/src/sizzle.js#L1404-L1412) because of a Safari bug).

  I’ve implemented a low-level `SelectAdapter.prototype.current` for this special case, it’s a bit faster but I guess the gain is only noticeable in my rather contrived performance stress test with 20,000 options.
2. When using ArrayAdapter, I understand the DOM has to be created upon the first Select2 opening. Some micro-optimizations would be possible there, but I guess the performance boils down to DOM element creation (`innerHTML`) and `$.data` being the bottleneck. 

  However, is it necessary to recreate the DOM with every subsequent opening?
  ```javascript
  Results.prototype.bind = function (container, $container) {
    // …
    container.on('results:all', function (params) {
      self.clear();
      self.append(params.data);
  ```
  The clear() operation is very costly, recreating all options too. AFAICS, recreating is currently necessary because

  - The select value might have changed
  - Templates might use the `selected` state, even return `null`

  Is it possible to implement a cache so the DOM doesn’t have to be created from scratch with each opening if the value hasn’t changed?
3. Three calls to `positionDropdown`. One in the beginning, two after DOM creation. This is costly because it involves synchronous layouting of all options due to `jQuery#position`, `position`, `outerHeight` calls.
4. `Results.prototype.setClasses` is relatively fast but could benefit from some optimization since it’s does heavy DOM operations (mostly `setAttribute` calls). For example, I guess the list `var $selected = $options.filter('[aria-selected=true]');` could be aggregated in the loop before, saving the DOM query.

In my real-world app, I have up to 1000 options with a fairly complex `templateResult`. I’ve implemented the pagination solution mentioned above to speed up the rendering, showing 100 options in each batch.

**Comment by kevin-brown** [comment]
I've been considering creating an adapter or decorator that enables pagination on local data sets, similar to how it is handled with remote data sets.

> Hello! I’m profiling with this example with a lot of options

Thank you! It's always good to see profiling done by others, as I usually make some mistake when doing it. :)

> `$.data` being the bottleneck

I've actually looked into this, and it used to be around 15x worse. [Here's a writeup I did on why we use `$.data`.](http://blog.kevin-brown.com/select2/2014/12/15/jquery-js-performance.html#dont-use-data-unless-you-need-data--attributes)

> However, is it necessary to recreate the DOM with every subsequent opening?

Yes, specifically for the case of AJAX and other custom data adapters.

> Results.prototype.setClasses is relatively fast but could benefit from some optimization since it’s does heavy DOM operations

I'm always open to pull requests with optimizations, as long as it doesn't break anything.
